### PR TITLE
Marcuse/16673

### PR DESCRIPTION
**Pull request by ifesdjeen**

### CODE REVIEW

File: src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java [hunk 1]
```diff
@@ -457,4 +433,48 @@ protected void validateExpectedOptions() throws ConfigurationException
                 throw new ConfigurationException(String.format("Unrecognized strategy option {%s} passed to %s for keyspace %s", key, getClass().getSimpleName(), keyspaceName));
         }
     }
+
+    static class ReplicaCache<K, V>
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java:443]: It looks like we can slightly improve this by introducing a versioned replicas map (i.e. a class that would hold a final cached version, and final map of replicas), that would be held in an atomic reference. Then, we don't need so synchronise on `this`, and can just create and return a new version of cached replicas.

File: src/java/org/apache/cassandra/locator/TokenMetadata.java [hunk 2]
```diff
@@ -656,6 +662,23 @@ public TokenMetadata cloneOnlyTokenMap()
         }
     }
 
+    private TokenMetadata cloneOnlyTokenMapWithRingVersion()
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/locator/TokenMetadata.java]: Is there a reason why we wouldn't always want to preserve a ring version? We could then just add `ringVersion` to the map above.

File: src/java/org/apache/cassandra/locator/TokenMetadata.java [hunk 3]
```diff
@@ -129,12 +129,18 @@ public TokenMetadata(IEndpointSnitch snitch)
...
+    }
+
+    private TokenMetadata(BiMultiValMap<Token, InetAddressAndPort> tokenToEndpointMap, BiMap<InetAddressAndPort, UUID> endpointsMap, Topology topology, IPartitioner partitioner, long ringVersion)
     {
         this.tokenToEndpointMap = tokenToEndpointMap;
         this.topology = topology;
         this.partitioner = partitioner;
         endpointToHostIdMap = endpointsMap;
         sortedTokens = sortTokens();
+        this.ringVersion = ringVersion;
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/locator/TokenMetadata.java:143]: `cloneWithNewPartitioner` should probably copy the ringVersion too.

File: src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java [hunk 4]
```diff
@@ -457,4 +435,54 @@ protected void validateExpectedOptions() throws ConfigurationException
...
+
+        ReplicaHolder(long ringVersion, ImmutableMap<K, V> replicas)
+        {
+            this.ringVersion = ringVersion;
+            this.replicas = replicas;
+        }
+
+        ReplicaHolder<K, V> withReplica(K key, V value)
+        {
+            return new ReplicaHolder<>(ringVersion, ImmutableMap.<K, V>builder().putAll(replicas).put(key, value).build());
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java]: Might be more efficient to initialize with an expected size. Looks like the default is 4 which we would exceed.

`ImmutableMap.<>builderWithExpectedSize(replicas.size()+1).putAll(replicas).put(key, value).build()`

File: src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java [hunk 5]
```diff
@@ -457,4 +435,54 @@ protected void validateExpectedOptions() throws ConfigurationException
...
+        void put(long ringVersion, K keyToken, V endpoints)
+        {
+            maybeClear(ringVersion);
+            ReplicaHolder<K, V> current = cachedReplicas.get();
+            // if we have the same ringVersion, but already know about the keyToken the endpoints should be the same
+            if (current.ringVersion == ringVersion && !current.replicas.containsKey(keyToken))
+            {
+                ReplicaHolder<K, V> next = current.withReplica(keyToken, endpoints);
+                cachedReplicas.compareAndSet(current, next);
+            }
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java]: Why use an `ImmutableMap` here with the cost of rebuilding it for all tokens once invalidated? The semantic is allowing writes at most once for a key, so why not stay with the lock-free `NonBlockingHashMap` and just use `putIfAbsent`?

```
if (current.ringVersion == ringVersion)
{
    current.replicas.putIfAbsent(keyToken, endpoints);
}
```

I ended up trying it out and pushed a version to https://github.com/jonmeredith/cassandra/tree/marcuse/16673 that makes the change and only bothers updating the hash if the ring version was current at the time of receiving it.

**Review by jonmeredith — COMMENTED** [review]
Looks like this fixes the issue nicely, but I wanted to discuss whether ImmutableMap was the right way to go.  I pushed up an alternative implementation to my repo.

**Review by jonmeredith — APPROVED** [review]
+1 

I think there's still a pending nit about cloneWithNewPartitioner not copying the ring version, and up to you if you'd like to add the stress test as well https://github.com/jonmeredith/cassandra/commit/349694482456b0bdbbd33cf2076b4c410a77cf78 (though the imports look like they need a cleanup)

**Pull request closed without merging**